pkg/cmd/kind/test/internal/testcmd/testcmd.go

```diff
@@ -0,0 +1,372 @@
+	config := &v1alpha4.Cluster{
+		Networking: v1alpha4.Networking{
+			IPFamily: v1alpha4.ClusterIPFamily(flags.IPFamily),
+		},
+		Nodes: []v1alpha4.Node{
+			{
+				Role:  v1alpha4.ControlPlaneRole,
+				Image: flags.ImageName,
+			},
+		},
+	}
+	// add e2e workers
+	numNodesForE2E := 2
+	for i := 0; i < numNodesForE2E; i++ {
+		config.Nodes = append(config.Nodes, v1alpha4.Node{
+			Role:  v1alpha4.WorkerRole,
+			Image: flags.ImageName,
+		})
+	}
+
+	// create the cluster
+	logger.V(0).Infof("Creating cluster %q ...\n", flags.Name)
+	if err := provider.Create(
+		flags.Name,
+		cluster.CreateWithV1Alpha4Config(config),
+		cluster.CreateWithRetain(true),
+		cluster.CreateWithWaitForReady(time.Minute),
+		cluster.CreateWithKubeconfigPath(flags.kubeconfigPath()),
+		cluster.CreateWithDisplayUsage(false),
+		cluster.CreateWithDisplaySalutation(false),
+	); err != nil {
+		if errs := errors.Errors(err); errs != nil {
+			for _, problem := range errs {
+				logger.Errorf("%v", problem)
+			}
+			return errors.New("aborting due to invalid configuration")
+		}
+		return errors.Wrap(err, "failed to create cluster")
+	}
+
+	// In Kubernetes CI we don't have real IPv6 connectivity so we need to
+	// employ some work arounds when testing IPv6 clusters
+	// We do this all environments for consistency.
+	if config.Networking.IPFamily == v1alpha4.IPv6Family {
+		if err := enableIPv6Workarounds(provider, flags.Name); err != nil {
+			return err
+		}
+	}
+
+	// run tests
+	skip := details.Skip
+	if details.Parallel {
+		skip = strings.Join([]string{`\[Serial\]`, skip}, "|")
+	}
+	cmd := exec.Command(
+		"./hack/ginkgo-e2e.sh",
+		// don't use the horrid e2e.test """provider""" concept
+		"--provider=skeleton",
+		// need to inform tests of worker count
+		"--num-nodes", strconv.Itoa(numNodesForE2E),
+		// write test results into artifacts
+		"--report-dir", flags.Artifacts,
+		// we do our own log dump
+		"--disable-log-dump=true",
+		// plumb ginkgo options
+		"--ginkgo.focus", details.Focus,
+		"--ginkgo.skip", skip,
+	).SetEnv(
+		// setting this env prevents ginkgo e2e from trying to run provider setup
+		"KUBERNETES_CONFORMANCE_TEST=y",
+		// setting these is required to make RuntimeClass tests work ... :/
+		"KUBE_CONTAINER_RUNTIME=remote",
+		"KUBE_CONTAINER_RUNTIME_ENDPOINT=unix:///run/containerd/containerd.sock",
+		"KUBE_CONTAINER_RUNTIME_NAME=containerd",
+		// ensure the test uses our kubeconfig
+		"KUBECONFIG="+flags.kubeconfigPath(),
+		"GINKGO_PARALLEL="+boolToYorN(details.Parallel),
+	)
+	return exec.InheritOutput(cmd).Run()
+}
+
+func boolToYorN(y bool) string {
+	if y {
+		return "y"
+	}
+	return "n"
+}
+
+func enableIPv6Workarounds(provider *cluster.Provider, name string) error {
+	// get a control plane node to use run kubectl from
+	nodes, err := provider.ListInternalNodes(name)
+	if err != nil {
+		return errors.Wrap(err, "failed to get nodes for coreDNS IPv6 CI fixes")
+	}
+	nodes, err = nodeutils.ControlPlaneNodes(nodes)
```

Comment:
nit, a different name to the assigned variable to distinguish from the previous one?

Reply:
that's anti-helpful here. we don't want to retain and accidentally use the unfiltered node list.
we could break this out into a helper method but it's just a few lines so that's not super helpful

Reply:
nah, no problem at all, is just there are at least 1 or 2 places more doing this that way, that's why I added the nit  😄